File: compiler/bench/crc32.fs
\ expected: 0
\ CRC32 checksum - compute CRC32 over many iterations

256 constant TABLE-SIZE
create crc-table TABLE-SIZE cells allot

: init-table ( -- )
  TABLE-SIZE 0 do
    i
    8 0 do
      dup 1 and if
        1 rshift $EDB88320 xor
      else
        1 rshift
      then
    loop
    crc-table i cells + !
  loop ;

: crc32-byte ( crc byte -- crc' )
  over xor $FF and cells crc-table + @
  swap 8 rshift xor ;

: crc32 ( addr len -- crc )
  $FFFFFFFF rot rot
  0 do
    over i + c@ crc32-byte
  loop
  nip $FFFFFFFF xor ;

create test-data 256 allot
: init-data ( -- )
  256 0 do i test-data i + c! loop ;

: main
  init-table
  init-data
  0
  10000 0 do
    test-data 256 crc32
    xor
  loop
  . cr ;

File: compiler/bench/crc32.c
// expected: 0
// CRC32 checksum - compute CRC32 over many iterations
#include <stdio.h>
#include <stdint.h>

uint32_t crc_table[256];
unsigned char test_data[256];

void init_table(void) {
    for (int i = 0; i < 256; i++) {
        uint32_t crc = i;
        for (int j = 0; j < 8; j++) {
            if (crc & 1)
                crc = (crc >> 1) ^ 0xEDB88320;
            else
                crc >>= 1;
        }
        crc_table[i] = crc;
    }
}

uint32_t crc32(unsigned char *data, int len) {
    uint32_t crc = 0xFFFFFFFF;
    for (int i = 0; i < len; i++) {
        crc = crc_table[(crc ^ data[i]) & 0xFF] ^ (crc >> 8);
    }
    return crc ^ 0xFFFFFFFF;
}

int main(void) {
    init_table();
    for (int i = 0; i < 256; i++) test_data[i] = i;
    uint32_t result = 0;
    for (int i = 0; i < 10000; i++) {
        result ^= crc32(test_data, 256);
    }
    printf("%u\n", result);
    return 0;
}
